Source/GSCSTableau.m

```diff
@@ -0,0 +1,700 @@
+  if ([variable isExternal])
+    {
+      [_externalRows setObject: expression forKey: variable];
+    }
+  [self addTermVariablesForExpression: expression variable: variable];
+}
+
+- (void) addTermVariablesForExpression: (GSCSLinearExpression *)expression
+                              variable: (GSCSVariable *)variable
+{
+  NSArray *termVariables = [expression termVariables];
+  FOR_IN(GSCSVariable *, expressionTermVariable, termVariables)
+    [self addMappingFromExpressionVariable: expressionTermVariable
+                             toRowVariable: variable];
+    if ([expressionTermVariable isExternal])
+      {
+        [_externalParametricVariables addObject: expressionTermVariable];
+      }
+  END_FOR_IN(termVariables);
+}
+
+- (void) addMappingFromExpressionVariable: (GSCSVariable *)columnVariable
+                            toRowVariable: (GSCSVariable *)rowVariable
+{
+  NSMutableSet *columnSet = [_columns objectForKey: columnVariable];
+  if (columnSet == nil)
+    {
+      columnSet = [NSMutableSet set];
+      [_columns setObject: columnSet forKey: columnVariable];
+    }
+  [columnSet addObject: rowVariable];
+}
+
+- (void) removeMappingFromExpressionVariable: (GSCSVariable *)columnVariable
+                               toRowVariable: (GSCSVariable *)rowVariable
+{
+  NSMutableSet *columnSet = [_columns objectForKey: columnVariable];
+  if (columnSet == nil)
+    {
+      return;
+    }
+  [columnSet removeObject: rowVariable];
+}
+
+- (void) removeColumn: (GSCSVariable *)variable
+{
+  NSSet *crows = [_columns objectForKey: variable];
+  if (_rows != nil)
+    {
+      FOR_IN(GSCSVariable *, clv, crows)
+        GSCSLinearExpression *expression = [_rows objectForKey: clv];
+        [expression removeVariable: variable];
+      END_FOR_IN(crows);
+      [_columns removeObjectForKey: variable];
+    }
+
+  if ([variable isExternal])
+    {
+      [_externalRows removeObjectForKey: variable];
+    }
+}
+
+- (void) removeRowForVariable: (GSCSVariable *)variable
+{
+  GSCSLinearExpression *expression = [_rows objectForKey: variable];
+  if (expression == nil)
+    {
+      NSException *missingExpressionException = [NSException
+        exceptionWithName: NSInvalidArgumentException
+                   reason: @"No expression exists for the provided variable"
+                 userInfo: nil];
+      [missingExpressionException raise];
+      return;
+    }
+  [_rows removeObjectForKey: variable];
+  if ([variable isExternal])
+    {
+      [_externalRows removeObjectForKey: variable];
+    }
+  NSArray *expressionTermVariables = [expression termVariables];
+  FOR_IN(GSCSVariable *, expressionTermVariable, expressionTermVariables)
+    [self removeMappingFromExpressionVariable: expressionTermVariable
+                                toRowVariable: variable];
+    if ([expressionTermVariable isExternal])
+      {
+        [_externalParametricVariables addObject: expressionTermVariable];
```

Comment:
This is correct, if the expression term variable is external it needs to be solved for.

Reply:
Could you please explain this a bit more? Perhaps this is easier for me to accept if you call `recordUpdatedVariable:` here.

Reply:
Honestly it's been a while since I wrote the code and I can't recall it's purpose. I have taken a look at the original implementation found here https://sourceforge.net/projects/cassowary/ and I can't find the check either. I have removed it.